support C++20 Modules, implement compilation

PikachuHyA opened the pull request:
I split the XXL PR https://github.com/bazelbuild/bazel/pull/19940 into several small patches.
This is the fourth patch of Support C++20 Modules, I modify the `CppCompileAction` and add related context, inputs and outputs.
- context: `.pcm` files and `.CXXModules.json` files
- inputs: `.pcm` files, `.modmap` file and `.modmap.input` file
- outputs: `.pcm` files and `.CXXModules.json` files

the main different in `CppCompileAction` is `computeUsedCpp20Modules`.

The restart mechanism is utilized; if the necessary C++20 Modules files are not ready during compilation, the current compilation will exit and wait for an appropriate time to recompile.

Note: To make the code review convenient, I keep only one commit. Due to patch dependencies, it may build and test failed.
I will update the patch after all the prerequisite patches have been merged. At that time, I will then troubleshoot the build and test failures.

peakschris commented:
@katre @lberki I'm not sure if you got the notification, but ping... If this PR [4/5] can be merged, we can start using C++ modules with bazel... in eager anticipation :-)

katre commented:
Sorry, I didn't notice this because I don't know anything about C++ modules. Re-assigning to @comius, apologies for not noticing this earlier.

PikachuHyA commented:
I'm testing this patch. If this patch ready, I will remove the `[WIP]` prefix.

peakschris commented:
@comius @trybka, friendly ping!

comius commented:
Hey, this is quite a large PR and will take me some time to review it completely. Thanks for your patience.

PikachuHyA commented:
> Hey, this is quite a large PR and will take me some time to review it completely.

Please also review https://github.com/bazelbuild/bazel/pull/22744, which contains a small patch split from this one.

PikachuHyA commented:
@comius kind ping

peakschris commented:
Hi @comius @trybka, is there any chance this feature could make it in soon? We are keenly awaiting it in one of the bazel 8 releases

PikachuHyA commented:
@comius ping

ChuanqiXu9 commented:
@comius @trybka can we expect to see this before 2025?

comius commented:
> @comius @trybka can we expect to see this before 2025?

Hey, yes. I was under a very heavy load for Bazel 8 release (removal of all the ruleset except C++). It's just winding down, so expect more going on in the following weeks.

PikachuHyA commented:
@comius @trybka ping

ChuanqiXu9 commented:
> > @comius @trybka can we expect to see this before 2025?
> 
> Hey, yes. I was under a very heavy load for Bazel 8 release (removal of all the ruleset except C++). It's just winding down, so expect more going on in the following weeks.

Not a push, but it's Dec 20.

tintor commented:
@comius @trybka It is January 2025. Many people are waiting for this PR to land.

peakschris commented:
@comius @trybka we are eagerly awaiting support for modules in Bazel. If you don't have time to review (this PR has now been waiting for 4 months), is there some other way it can progress?

sachnk commented:
@trybka joining others in voicing excitement in getting c++ module support into bazel. Anything we can do to help get this merged? Any rough ETA so we can plan accordingly?

Much appreciated!

PikachuHyA commented:
> I conferred with @comius and I'm going to take point. As I understand it, it doesn't merge cleanly in our repo so I anticipate some back and forth to get it working.

ping @trybka

trybka commented on src/main/java/com/google/devtools/build/lib/rules/cpp/CcCompilationHelper.java:
```diff
@@ -1001,13 +1021,287 @@ private ImmutableSet<Artifact> getSourceArtifactsByType(
+                  CompileBuildVariables.CPP_MODULE_MODMAP_FILE.getVariableName(),
+                  modmapFile.getExecPathString())
+              .build();
+      createSourceActionHelper(
```
nit:
```
error: [CheckReturnValue] The result of `createSourceActionHelper(...)` must be used
      createSourceActionHelper(
                              ^
  If you really don't want to use the result, then assign it to a variable: `var unused = ...`.
  
  If callers of `createSourceActionHelper(...)` shouldn't be required to use its result, then annotate it with `@CanIgnoreReturnValue`.
  ```

PikachuHyA replied:
annotate `createSourceActionHelper` with @CanIgnoreReturnValue

trybka commented on src/main/java/com/google/devtools/build/lib/rules/cpp/CcCompilationHelper.java:
```diff
@@ -1001,13 +1021,287 @@ private ImmutableSet<Artifact> getSourceArtifactsByType(
+                  CompileBuildVariables.CPP_MODULE_MODMAP_FILE.getVariableName(),
+                  modmapFile.getExecPathString())
+              .build();
+      createSourceActionHelper(
```
nit:

```
error: [CheckReturnValue] The result of `createSourceActionHelper(...)` must be used
      createSourceActionHelper(
                              ^
  If you really don't want to use the result, then assign it to a variable: `var unused = ...`.
  
  If callers of `createSourceActionHelper(...)` shouldn't be required to use its result, then annotate it with `@CanIgnoreReturnValue`.
  ```

trybka reviewed (commented):
I have one test I'm trying to wrap up -- since we use a different semantics.bzl internally, one of the new test cases is currently failing.

I left some comments from our code checker.

I did a rebase and resolved merges this morning. 

In the interim, I have built a new `bazel` with these changes and am running a global presubmit internally to see if that uncovers any additional issues.

PikachuHyA commented:
Rebased due to windows CI failed.

trybka commented:
Internal checks mostly look good. AFAICT no adverse breakages, but I'm re-running flaky failing tests anyway.

I don't have a good solution yet for the failing test. I'm wondering if it would be possible to gate the tests that require cpp20 modules support on some other bit that signals that we have the requisite tools available? (Maybe it would be sufficient to try to configure the action once and fail gracefully).

I can imagine also wanting something in the crosstool itself (e.g. `feature(name='supports_cpp20_modules')`) for the non-{clang,msvc-cl,gcc} compilers or toolchains not yet willing to adopt support.

Here's the current failure:

Basically because we don't intend to support C++20 modules immediately, we had stubbed out the tools provided in semantics.bzl -- e.g. here: https://cs.opensource.google/bazel/bazel/+/master:src/main/starlark/builtins_bzl/common/cc/semantics.bzl;l=137-149;drc=f777a3c33151b5a3ae2295f1ba6ea94653840665

Ours looks like this:
```
semantics = struct(
   ...
   cpp_modules_tools = lambda: {},
   ...
)
```

With this PR, tests like `CppModulesConfiguredTargetTest.testSameModuleInterfacesFileInCcLibraryTwice` fail because those tools are not provided, e.g. with:

```
java.lang.RuntimeException: Unrecoverable error while evaluating node 'ConfiguredTargetKey{label=<somelabel>, config=BuildConfigurationKey[<somekey>]}' (requested by nodes 'ConfiguredTargetKey{label=<someotherlabel>, config=BuildConfigurationKey[<somekey>]}')
...
Caused by: net.starlark.java.eval.Starlark$UncheckedEvalException: NullPointerException thrown during Starlark evaluation (<somelabel>)
	at <starlark>.compile(<builtin>:0)
	at <starlark>._compile(/virtual_builtins_bzl/common/cc/cc_common.bzl:755)
	at <starlark>._cc_library_impl(/virtual_builtins_bzl/common/cc/cc_library.bzl:56)
Caused by: java.lang.NullPointerException: the tool 'aggregate-ddi' cannot be null
	at com.google.common.base.Preconditions.checkNotNull(Preconditions.java:988)
	at com.google.devtools.build.lib.rules.cpp.CcCompilationHelper.createAggDdiAction(CcCompilationHelper.java:1852)
	at com.google.devtools.build.lib.rules.cpp.CcCompilationHelper.createCcCompileActionsWithCpp20ModuleHelper(CcCompilationHelper.java:1057)
	at com.google.devtools.build.lib.rules.cpp.CcCompilationHelper.createCcCompileActionsWithCpp20Module(CcCompilationHelper.java:987)
	at com.google.devtools.build.lib.rules.cpp.CcCompilationHelper.compile(CcCompilationHelper.java:835)
	at com.google.devtools.build.lib.rules.cpp.CcModule.compile(CcModule.java:2255)
```

PikachuHyA commented:
> With this PR, tests like `CppModulesConfiguredTargetTest.testSameModuleInterfacesFileInCcLibraryTwice` fail because those tools are not provided

hi @trybka , thank you for sharing the detailed failure messages.

Would it be possible for us to temporarily disable the failing tests in order to quickly merge this patch?

PikachuHyA commented:
> Yes that seems reasonable. If you want to do that I can merge your changes, otherwise I'll do something tomorrow.

hi @trybka , I have removed the `CppModulesConfiguredTargetTest` . please merge the patch.

PikachuHyA commented:
I believe the CI failure is not related to this patch.

fmeum commented:
The failures are known, I retried the runs.

PikachuHyA commented:
> Can you try rebasing? 

Rebased.

trybka commented:
Mailed for review internally -- local presubmits all green. Global presubmit hit a snag and didn't run. Scheduled another one for this evening, will try to check it early in the AM.

trybka commented:
Global presubmits are mostly green--just reviewing the failures and ensuring they are flakes.

Reviewer has provided feedback. I'm going to try to address as much as possible as local patches rather than incur more back and forth, @PikachuHyA. I'll post if I need any assistance resolving the additional comments.

PikachuHyA commented:
> Global presubmits are mostly green--just reviewing the failures and ensuring they are flakes.
> 
> Reviewer has provided feedback. I'm going to try to address as much as possible as local patches rather than incur more back and forth, @PikachuHyA. I'll post if I need any assistance resolving the additional comments.

ping @trybka

trybka reviewed (commented):
Question from the review inline.

Working on other comments and some fixes. I expect another round.

PikachuHyA commented on src/main/java/com/google/devtools/build/lib/rules/cpp/CppCompileAction.java:
```diff
@@ -1552,6 +1619,65 @@ private boolean shouldParseShowIncludes() {
+    }
+    for (String modulePath : modulePathList) {
+      if (moduleFileMap.containsKey(modulePath)) {
+        usedModules.add(moduleFileMap.get(modulePath));
```
Yes, we now skip the error here. The modmap file is generated by the [generate-modmap](https://github.com/bazelbuild/bazel/blob/master/tools/cpp/modules_tools/BUILD#L72) tool. We have already checked the modules used by the current translation unit. For details, see 

https://github.com/bazelbuild/bazel/blob/master/tools/cpp/modules_tools/generate-modmap/generate-modmap.cc#L94-L101

Therefore, the `modulePath` is generally included in the `moduleFileMap`.

However, if the `modulePath` is not found in the `moduleFileMap` due to an unexpected issue (such as a bug), the compiler will raise an error like `xxx not found` during the compilation of the current TU. Thus, the missing case does not affect the final result; the only difference is whether the error message originates from the build system (Bazel) or the compiler.

Do you have a preference between these two approaches? It would be fine if we report it here.

Perhaps adding an assertion here would be sufficient.

PikachuHyA commented:
@trybka ping

trybka commented:
> @trybka ping

Still waiting on internal review. I believe my reviewers are OOO for the last week.

PikachuHyA commented:
@trybka ping

trybka commented:
@comius had mentioned that they had left a comment asking for clarification on a piece of this, but I can't find the question they are referring to. I have been waiting on resolution of that, as well as other concerns that were raised, possibly to do with SkyGraph?

cc: @comius @pzembrod for visibility / clarification

comius commented:
> @comius had mentioned that they had left a comment asking for clarification on a piece of this, but I can't find the question they are referring to. I have been waiting on resolution of that, as well as other concerns that were raised, possibly to do with SkyGraph?
> 
> cc: @comius @pzembrod for visibility / clarification

Uff, sorry, I just figured out that my review comments were Pending and not submitted.

PikachuHyA commented:
@trybka @comius any updates on this patch?

PikachuHyA commented:
@trybka ping

trybka commented:
@comius to follow up

comius commented on src/main/java/com/google/devtools/build/lib/rules/cpp/CppCompileAction.java:
```diff
@@ -1552,6 +1619,65 @@ private boolean shouldParseShowIncludes() {
+      }
+    }
+    // use restart mechanism
+    var env = actionExecutionContext.getEnvironmentForDiscoveringInputs();
+    var skyKeys = Collections2.transform(usedModules, DerivedArtifact::getGeneratingActionKey);
+    var actionExecutionValues = env.getValuesAndExceptions(skyKeys);
+    if (env.valuesMissing()) {
+      return null;
+    }
+    for (DerivedArtifact module : usedModules) {
+      Preconditions.checkState(
+          module.isFileType(CppFileTypes.CPP_MODULE), "Non-module? %s", module);
+      var skyValue = actionExecutionValues.get(module.getGeneratingActionKey());
+      if (skyValue == null) {
+        return null;
+      }
+    }
```
> If Skyframe has alternative methods to ensure that these dependent module files exist at the time of compiling the TU, then this part of the code can be removed.

When using regular inputs or even discovered inputs, Skyframe will already do everything for you. Manual Skyframe triggering is really not needed in your case.

> I also tested the implementation after removing this code, and it builds successfully in the unit tests, integration tests, and the async_simple project.

I suspect that this code is unnecessarily complex. 

I can't provide very good guidance on how to simplify the code. I'm also concerned that I'll potentially spend more time reviewing the new approach as I'd spend simplifying it.

PikachuHyA replied:
- Rebased to resolve conflicts.
- Removed unnecessary manual triggering of Skyframe.

fmeum commented:
@PikachuHyA I'll take over the review. Just know that I'm not a Bazel team member, so even when I approve, it doesn't guarantee that the PR will be accepted as is. Are you okay with me sending a PR against your branch with a few suggested improvements? That would be easier than going through inline suggestions in the GitHub UI.

fmeum commented on src/main/java/com/google/devtools/build/lib/rules/cpp/CcCompilationHelper.java:
```diff
@@ -950,13 +970,287 @@ private ImmutableSet<Artifact> getSourceArtifactsByType(
+      builder.setModmapInputFile(modmapInputFile);
+
+      PathFragment ccRelativeName = sourceArtifact.getRootRelativePath();
+      // TODO(plf): Continue removing CLIF logic from C++. Follow up changes would include
```
We can drop all the CLIF stuff since we know that our source is a module interface file.

fmeum reviewed (commented):
I sent this PR to your fork with suggested improvements: https://github.com/PikachuHyA/bazel/pull/8

PikachuHyA commented:
> @PikachuHyA I'll take over the review. Just know that I'm not a Bazel team member, so even when I approve, it doesn't guarantee that the PR will be accepted as is. 

@fmeum Thank you very much for offering to take over the review! I really appreciate your time and help.

> Are you okay with me sending a PR against your branch with a few suggested improvements? That would be easier than going through inline suggestions in the GitHub UI.

Yes, I’m perfectly fine with you opening a PR against my branch—that will indeed be more convenient than going through a long list of inline comments.

I’ve gone through your suggestions and agree with them in general. I just have one point I’d like to double-check:

The ddi files in `ImmutableList.copyOf(sourceToDdiFileMap.values())` is expected to stay in a 1-to-1 correspondence with the pcm files in `directModuleFiles (NestedSet with Order.STABLE_ORDER)`. I’ve tested several cases and they all match, but could you please confirm that my understanding is correct?

Thanks again for your review.

PikachuHyA commented on src/main/java/com/google/devtools/build/lib/rules/cpp/CppCompileActionBuilder.java:
```diff
@@ -355,6 +363,19 @@ NestedSet<Artifact> buildMandatoryInputs() throws EvalException {
       realMandatoryInputsBuilder.addTransitive(ccCompilationContext.getDeclaredIncludeSrcs());
       realMandatoryInputsBuilder.addTransitive(additionalPrunableHeaders);
     }
+    if (CppActionNames.CPP_MODULE_DEPS_SCANNING.equals(actionName)) {
```
I agree. At the moment the only action that produces a `.modmap` file is `c++20-module-compile`, so `modmapFile` and `modmapInputFile` will either both be non-null or both be null. We can therefore drop the layered conditionals and just add them whenever they’re non-null. I’ll update the code to reflect that.

PikachuHyA commented on src/main/java/com/google/devtools/build/lib/rules/cpp/CppCompileAction.java:
```diff
@@ -383,15 +408,34 @@ private static ImmutableSet<Artifact> collectOutputs(
     outputs.addAll(additionalOutputs);
     if (dotdFile != null) {
-      outputs.add(dotdFile);
+      if (featureConfiguration.isEnabled(CppRuleClasses.CPP_MODULES)) {
```
You’re right—those changes aren’t necessary in this PR.

This PR adds support for C++20 modules in the one-phase compilation pipeline, which does not create a separate `c++20-module-codegen` action. Consequently, we never set a `dotdFile` explicitly. The only `dotdFile` involved is the one that `c++20-module-compile` automatically receives via `builder.setOutputs(...)` inside `createSourceActionHelper`.

The explicit `dotdFile` logic in `CppCompileAction.java` is needed only for the two-phase pipeline in https://github.com/bazelbuild/bazel/pull/22555 , where `c++20-module-codegen` reuses the `dotdFile` produced by `c++20-module-compile` to prune dependencies.

To keep this PR clean and focused, I’ll remove the extra branching from the action.

PikachuHyA commented on src/main/java/com/google/devtools/build/lib/rules/cpp/CppCompileAction.java:
```diff
@@ -753,7 +805,7 @@ public NestedSet<Artifact> getDiscoveredModules() {
 
   /** Returns the path where the compiler should put the discovered dependency information. */
   public Artifact getDotdFile() {
-    return compileCommandLine.getDotdFile();
```
Understood—I’ll revert this change so this PR keeps the original semantics. That tweak is only needed for the two-phase compilation work discussed in https://github.com/bazelbuild/bazel/pull/22553#discussion_r2179213785 .

PikachuHyA commented on src/test/java/com/google/devtools/build/lib/rules/cpp/CppModulesConfiguredTargetTest.java:
```diff
@@ -1,271 +0,0 @@
-// Copyright 2024 The Bazel Authors. All rights reserved.
```
> Could you also add an integration test for single-pass module compilation to cc_integration_test.sh?

Regarding an integration test in `cc_integration_test.sh`: single-pass module compilation requires a toolchain whose compiler is built with C++20-modules support. I’m not sure whether the compilers available on all CI shards have this capability.

If the CI toolchains do support C++20 Modules, I’ll gladly add an end-to-end test.

fmeum replied:
@meteorcloudy Do you know which CI runners have a modern enough clang for this?

fmeum commented:
@PikachuHyA I meant to respond to all comments with open questions and also send a fix for the comment you had on my PR - let me know if you need anything else from me at this time.

PikachuHyA commented on src/main/java/com/google/devtools/build/lib/rules/cpp/CcCompilationHelper.java:
```diff
@@ -950,13 +970,287 @@ private ImmutableSet<Artifact> getSourceArtifactsByType(
+              configuration);
+      var ddiFile = ddiFileMap.get(sourceArtifact);
+      Preconditions.checkNotNull(ddiFile);
+      createGenModmapAction(ddiFile, modulesInfoFile, modmapFile, modmapInputFile);
```
see https://github.com/bazelbuild/bazel/pull/26471

PikachuHyA replied:
Done

https://github.com/bazelbuild/bazel/commit/7638d5360e173ac5f043adc2d55893644cc6f4af

PikachuHyA commented on src/main/java/com/google/devtools/build/lib/rules/cpp/CppCompileAction.java:
```diff
@@ -966,7 +1018,12 @@ public ImmutableMap<String, String> getEffectiveEnvironment(
       environment.put("PWD", "/proc/self/cwd");
     }
-
+    if (Objects.equals(CppActionNames.CPP_MODULE_DEPS_SCANNING, actionName)) {
```
Thanks for the guidance! I found a Starlark example that uses `env_entry` in rules_cc:

https://github.com/bazelbuild/rules_cc/blob/fe41fc4ea219c9d3680ee536bba6681f3baf838e/cc/private/toolchain/unix_cc_toolchain_config.bzl#L454-L460

Because the C++ toolchain definitions now live in rules_cc, I opened a PR there to add the environment entry for `clang-scan-deps`.


see https://github.com/bazelbuild/rules_cc/pull/428

PikachuHyA replied:
Done

https://github.com/bazelbuild/rules_cc/commit/bdd93e82a0bf2a9f9f9a221cbb9136fe325e95b6

PikachuHyA commented on src/main/java/com/google/devtools/build/lib/rules/cpp/CcCompilationHelper.java:
```diff
@@ -950,13 +970,287 @@ private ImmutableSet<Artifact> getSourceArtifactsByType(
+    for (CppSource source : moduleInterfaceSources.values()) {
+      Artifact sourceArtifact = source.getSource();
+      String outputName = outputNameMap.get(sourceArtifact);
+      ArtifactCategory outputCategory = ArtifactCategory.CPP_MODULE;
```
You’re absolutely right—using multiple ArtifactCategories here was incorrect.  
I’ll update this PR to keep only `ArtifactCategory.CPP_MODULE` and remove the compiler-specific checks from the Java code; the file extension is now selected exclusively through `artifact_name_pattern` in the toolchain.

The corresponding toolchain update is in rules_cc PR, see https://github.com/bazelbuild/rules_cc/pull/429

Thanks for catching this!

PikachuHyA commented on src/main/java/com/google/devtools/build/lib/rules/cpp/CppCompileAction.java:
```diff
@@ -1540,6 +1602,13 @@ public ActionResult execute(ActionExecutionContext actionExecutionContext)
             pathMapper);
     dotDContents = null; // Garbage collect in-memory .d contents.
 
+    if (usedCpp20Modules != null) {
```
Thanks for raising this. Unfortunately I don’t have access to a Windows machine right now, so I can’t verify the `/showIncludes` path directly. 

Full MSVC C++20 Modules support will need additional work anyway, and we can revisit this code path then. If you do see any issues on a Windows build, please let me know and I’ll address them promptly.

Thanks again for the thorough review!

fmeum commented:
@PikachuHyA An existing test fails on all OSes, could you take a look at that?

I will do a final review pass when CI is green, then I'll hand over to the Bazel team.

PikachuHyA commented on src/main/java/com/google/devtools/build/lib/rules/cpp/CcCompilationHelper.java:
```diff
@@ -1487,6 +1768,108 @@ private void createHeaderAction(
+            .addExecPath(ddiFile)
+            .addExecPath(modulesInfoFile)
+            .addExecPath(modmapFile)
+            .addDynamicString(ccToolchain.getCompiler())
```
I tried to replace the explicit `--compiler` flag with a wrapper script, but I ran into a path-resolution issue: I don’t know how to reference the `@bazel_tools//tools/cpp:generate-modmap` binary from inside the wrapper at the moment it is executed.

Attempt 1 – repository-time template  
I create a shell script with `repository_ctx.template(...)`:

```sh
set -e
%{generate_modmap} $@ --compiler=%{compiler}
```

• `%{compiler}` is easy: we can read it from 

https://github.com/bazelbuild/rules_cc/blob/d068b730235538bf1469c90b50b3c99e3c2152ca/cc/private/toolchain/unix_cc_configure.bzl#L658-L662


• `%{generate_modmap}` is the problem: at repository-rule time I don’t have an
  `artifact` or path for the label `@bazel_tools//tools/cpp:generate-modmap`, so I
  can’t substitute it into the template. I tried with `resolve_labels`, but the path is error.

Attempt 2 – Starlark rule that writes the wrapper  

```python
def _impl(ctx):
    ctx.actions.write(
        output = ctx.outputs.output,,
        content = """
        set -ex
        {generate_modmap} $@ --compiler {compiler}
        """.format(
            generate_modmap = ctx.executable._generate_modmap.path,
            compiler = ctx.attr.compiler,
        ),
        is_executable = True,
    )
    return [DefaultInfo(
        executable = ctx.outputs.output,
        runfiles = ctx.runfiles(
            files = [ctx.executable._generate_modmap],
            collect_data = True
        ),
    )]

generate_modmap_wrapper = rule(
    implementation = _impl,
    attrs = {
        "compiler": attr.string(
            mandatory = True,
            doc = "The compiler to use.",
        ),
        "_generate_modmap": attr.label(
            default = "@bazel_tools//tools/cpp:generate-modmap",
            executable = True,
            cfg = "exec",
        ),
        "output": attr.output(
            mandatory = True,
            doc = "The output file."
        ),
    },
    executable = True,
)
```

The rule builds fine, but when the `CppGenModmap` action runs I get

```
bazel-out/k8-opt-exec-ST-d57f47055a04/bin/external/rules_cc++cc_configure_extension+local_config_cc/generate_modmap_wrapper.sh: line 3: external/bazel_tools/tools/cpp/modules_tools/generate-modmap: No such file or directory
```

Even though I listed the binary in `runfiles`, it isn’t present in the sandbox at runtime.

fmeum replied:
A simple alternative: Modify `generate-modmap` to check the suffix of its `argv[0]`, then have the toolchains `copy_file` with `allow_symlink = True` it to the appropriate name.

PikachuHyA replied:
@fmeum The `generate-modmap` binary isn’t placed in the sandbox. My guess is that this happens because `getGenerateModmap` returns an `Artifact` rather than a `DefaultInfo` (or any other provider that carries runfiles).

As a workaround, I added a new `generate_modmap_wrapper` attribute and obtain it with `getGenerateModmapWrapper`. I then explicitly add the original `generate-modmap` binary to the action’s inputs via `actions.addInput(genModmapTool)`, which succeeds in placing it in the sandbox.

```
    var genModmapTool = ccToolchain.getGenerateModmap();
    if (genModmapTool == null) {
      throw new EvalException("the 'generate_modmap' tool is not defined in the C++ toolchain");
    }
    var genModmapToolWrapper = ccToolchain.getGenerateModmapWrapper();
    if (genModmapToolWrapper == null) {
      throw new EvalException("the 'generate_modmap_wrapper' tool is not defined in the C++ toolchain");
    }
    var commandLine =
        CustomCommandLine.builder()
            .addExecPath(ddiFile)
            .addExecPath(modulesInfoFile)
            .addExecPath(modmapFile)
            .build();
    var genModmapAction =
        new SpawnAction.Builder()
            .setExecutable(genModmapToolWrapper)
            .addCommandLine(commandLine)
            .addInput(genModmapTool)
            .addInput(ddiFile)
```

see https://github.com/bazelbuild/bazel/pull/22553/commits/adc682bf3c387e94055b68f030a34f38727c3898#diff-403c46ec3075b8e9e6d490ce955db88dae2d457b8046608884039b18b10ab6cc

Is this the correct way to solve the `generate-modmap` issue, or is there a cleaner approach you’d recommend?

fmeum replied:
Could you drop the `_wrapper` suffix from the attribute? It doesn't really matter for the toolchain what we plug in there. I don't see another way to do this since no other attribute picks up runfiles.

fmeum replied:
Re 1. Whenever something C++-specific can live in rules_cc, it should :⁠-⁠)

PikachuHyA commented:
> @PikachuHyA What's the minimum version of clang required for an integration test? CI has Ubuntu 24.04 with default clang installed.


I realized I had misunderstood the CI configuration earlier—sorry for the confusion.

I reran the tests on Ubuntu 24.04, which ships with Clang 18 by default.

```
$ clang -v
Ubuntu clang version 18.1.3 (1ubuntu1)
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /usr/bin
Found candidate GCC installation: /usr/bin/../lib/gcc/x86_64-linux-gnu/13
Selected GCC installation: /usr/bin/../lib/gcc/x86_64-linux-gnu/13
Candidate multilib: .;@m64
Selected multilib: .;@m64
```

Because we introduced a wrapper for `generate_modmap` (see https://github.com/bazelbuild/bazel/pull/22553#discussion_r2202381024), we also need to update `rules_cc`. see  https://github.com/bazelbuild/rules_cc/pull/447.

Once the `generate_modmap` wrapper issue is fully resolved, a complete C++20 Modules integration test should pass without problems. 

If needed, I can migrate the tests from this repository— https://github.com/PikachuHyA/bazel_cxx20_module_test —into the Bazel integration test suite.

fmeum commented:
@PikachuHyA I sent https://github.com/bazelbuild/continuous-integration/pull/2358 to add clang-scan-deps to the docker images. You might have to improve the logic that locates the binary since its installed as `clang-scan-deps-N` for some `N` by default.

fmeum commented on src/main/starlark/builtins_bzl/common/cc/compile/compile.bzl:
```diff
@@ -436,6 +493,602 @@ def _should_provide_header_modules(
+        ),
+    )
+    if use_pic:
+        outputs.add_pic_modules_info_file(modules_info_file)
```
You could add a `pic` `boolean` parameter to these functions, which cuts their number in half and also makes the callsite more convenient. I know the existing functions don't do that, but we don't have to add to that mess :-) 

Force the parameter to be named, not positional, so that the call sites are easy to read.

fmeum commented on src/main/starlark/builtins_bzl/common/cc/compile/compile.bzl:
```diff
@@ -436,6 +493,602 @@ def _should_provide_header_modules(
+        _create_gen_modmap_action(
+            actions = actions,
+            cc_toolchain = cc_toolchain,
+            ddi_file = source_to_ddi_file_map.get(source_artifact),
```
Do we ever expect `source_artifact not in source_to_ddi_file_map`? If so, please add a comment explaining what happens in that case, if not, use `[...]` instead of `get` to make this an error.

fmeum commented on src/main/starlark/builtins_bzl/common/cc/compile/compile.bzl:
```diff
@@ -436,6 +493,602 @@ def _should_provide_header_modules(
+            generate_dwo = should_create_per_object_debug_info(feature_configuration, cpp_configuration),
+            bitcode_output = bitcode_output,
+            additional_build_variables = {
+                "cpp_module_output_file": module_file.path,
+                "cpp_module_modmap_file": modmap_file.path
```
```suggestion
                "cpp_module_output_file": module_file,
                "cpp_module_modmap_file": modmap_file,
```
should work just as well and support path mapping

fmeum commented on src/main/starlark/builtins_bzl/common/cc/compile/compile.bzl:
```diff
@@ -436,6 +493,602 @@ def _should_provide_header_modules(
+                    output_name = output_name_base,
+                ),
+            )
+            additional_build_variables["cpp_module_modmap_file"] = modmap_file.path
```
```suggestion
            additional_build_variables["cpp_module_modmap_file"] = modmap_file
```

fmeum commented on src/main/starlark/builtins_bzl/common/cc/compile/compile.bzl:
```diff
@@ -436,6 +493,602 @@ def _should_provide_header_modules(
+    """
+    output_name_prefix_dir = cc_internal.compute_output_name_prefix_dir(configuration = configuration, purpose = purpose)
+    output_name_map = _calculate_output_name_map_by_type(compilation_unit_sources | module_interfaces_sources, output_name_prefix_dir)
+    if generate_no_pic_action:
```
Avoid the duplicated long call via logic like this:
```starlark
use_pic_values = []
if generate_no_pic_action:
  use_pic_values.append(false)
if generate_pic_action:
  use_pic_values.append(true)

for use_pic in use_pic_values:
  ...
```

fmeum reviewed (commented):
Sorry for the busywork, but yes, I don't think we need https://github.com/bazelbuild/rules_cc/pull/447 anymore.

fmeum commented:
That's great! Let me know when CI is green (or you need help) and I'll do a final pass before passing this on to the Bazel team.

fmeum commented:
Just found https://clang.llvm.org/docs/StandardCPlusPlusModules.html#source-files-consistency - do we need `-fmodules-embed-all-files`?

ChuanqiXu9 commented:
> Just found https://clang.llvm.org/docs/StandardCPlusPlusModules.html#source-files-consistency - do we need `-fmodules-embed-all-files`?

In downstream, we used `-fmodules-embed-all-files` by default. If it helps the problem you mentioned, I think we should turn it on.

fmeum commented:
I'll take a look.

The absolute path leakage is fixed in LLVM 21, so we don't have to worry about that anymore. At some point we will want a hermiticity test, but we can wait until 21 is available more widely.

fmeum commented on src/test/shell/bazel/cc_integration_test.sh:
```diff
@@ -2117,4 +2117,87 @@ EOF
+EOF
+  cat > main.cc <<'EOF'
+import foo;
+import bar;
+
+void f() {
+  f_foo();
+  f_bar();
+}
+
+int main() {
+  f();
+  return 0;
+}
+EOF
+  cat > foo.cppm <<'EOF'
+export module foo;
+import base;
+
+export void f_foo() {
+  f_base();
+}
+EOF
+  cat > bar.cppm <<'EOF'
+export module bar;
+import base;
+
+export void f_bar() {
+  f_base();
+}
+EOF
+  cat > base.cppm <<'EOF'
+export module base;
+
+export void f_base() {
+}
+EOF
+
+  bazel build //:main --repo_env=CC=clang --experimental_cpp_modules || fail "Expected build C++20 Modules success with compiler 'clang'"
+}
```
I turned this into a reproducer for the self cycle and added a test to cover the true cycle you brought up (also a minor TODO on making `--cxxopt` work for module interfaces, but let's do that in a follow-up PR):

```suggestion
function test_cpp20_modules_with_clang() {
  type -P clang || return 0
  # Check if clang version is less than 17
  clang_version=$(clang --version | head -n1 | grep -oE '[0-9]+\.[0-9]+' | head -n1)
  if [[ -n "$clang_version" ]]; then
    major_version=$(echo "$clang_version" | cut -d. -f1)
    if [[ "$major_version" -lt 17 ]]; then
      return 0
    fi
  fi

  add_rules_cc "MODULE.bazel"
  add_to_bazelrc "common --repo_env=CC=clang"
  # TODO: Make it so that --cxxopt applies to module_interfaces as well.
  add_to_bazelrc "common --copt=-std=c++20"

  cat > BUILD.bazel <<'EOF'
load("@rules_cc//cc:defs.bzl", "cc_library", "cc_binary")

package(features = ["cpp_modules"])

cc_library(
  name = "base",
  module_interfaces = ["base.cppm"],
)
cc_library(
  name = "foo",
  module_interfaces = ["foo.cppm"],
  deps = [":base"]
)
cc_library(
  name = "bar",
  module_interfaces = ["bar.cppm"],
  deps = [":base"]
)
cc_binary(
  name = "main",
  srcs = ["main.cc"],
  deps = [":foo", ":bar"]
)
EOF
  cat > main.cc <<'EOF'
import foo;
import bar;

void f() {
  f_foo();
  f_bar();
}

int main() {
  f();
  return 0;
}
EOF
  cat > foo.cppm <<'EOF'
export module foo;
import base;

export void f_foo() {
  f_base();
}
EOF
  cat > bar.cppm <<'EOF'
export module bar;
import base;

export void f_bar() {
  f_base();
}
EOF
  cat > base.cppm <<'EOF'
export module base;

export void f_base() {
}
EOF

  bazel build //:main --experimental_cpp_modules --disk_cache=disk || fail "Expected build C++20 Modules success with compiler 'clang'"

  # Verify that the build can hit the cache without action cycles.
  bazel clean || fail "Expected clean success"
  bazel build //:main --experimental_cpp_modules --disk_cache=disk || fail "Expected build C++20 Modules success with compiler 'clang'"
}

function test_cpp20_modules_with_clang_no_cycle() {
  type -P clang || return 0
  # Check if clang version is less than 17
  clang_version=$(clang --version | head -n1 | grep -oE '[0-9]+\.[0-9]+' | head -n1)
  if [[ -n "$clang_version" ]]; then
    major_version=$(echo "$clang_version" | cut -d. -f1)
    if [[ "$major_version" -lt 17 ]]; then
      return 0
    fi
  fi

  add_rules_cc "MODULE.bazel"
  add_to_bazelrc "common --repo_env=CC=clang"
  # TODO: Make it so that --cxxopt applies to module_interfaces as well.
  add_to_bazelrc "common --copt=-std=c++20"

  cat > BUILD.bazel <<'EOF'
load("@rules_cc//cc:defs.bzl", "cc_library", "cc_binary")

package(features = ["cpp_modules"])

cc_library(
  name = "lib",
  module_interfaces = ["foo.cppm", "bar.cppm"],
)
EOF
  cat > foo.cppm <<'EOF'
export module foo;
import bar;
EOF
  cat > bar.cppm <<'EOF'
export module bar;
EOF

  bazel build //:lib --experimental_cpp_modules || fail "Expected build C++20 Modules success with compiler 'clang'"

  cat > foo.cppm <<'EOF'
export module foo;
EOF
  cat > bar.cppm <<'EOF'
export module bar;
import foo;
EOF

  bazel build //:lib --experimental_cpp_modules || fail "Expected build C++20 Modules success with compiler 'clang'"
}
```

fmeum commented on src/test/shell/bazel/cc_integration_test.sh:
```diff
@@ -2117,4 +2117,136 @@ EOF
+
+  # Verify that the build can hit the cache without action cycles.
+  bazel clean || fail "Expected clean success"
+  bazel build //:main --experimental_cpp_modules --disk_cache=disk || fail "Expected build C++20 Modules success with compiler 'clang'"
```
```suggestion
  bazel build //:main --experimental_cpp_modules --disk_cache=disk || fail "Expected build C++20 Modules success with compiler 'clang'"
  expect_log "20 disk cache hit"
```

adds some additional assurance.

I have an idea how to solve this, but it doesn't resolve the other cycle. I'll keep looking.

fmeum commented:
While there are a number of ways to solve the self-cycle, the true cycle is legitimately hard to fix. I have asked around for ideas.

@PikachuHyA How much worse would the user experience be if we only allowed a single module interface per target?

mathstuf commented:
> How much worse would the user experience be if we only allowed a single module interface per target?

Would it be allowed to have interface partitions? Note that "how big should modules be?" is an open question in C++ with modules; I don't know that forcing it at the build system level will allow appropriate exploration of the space to get a good answer.

PikachuHyA commented:
> > How much worse would the user experience be if we only allowed a single module interface per target?
> 
> Would it be allowed to have interface partitions? Note that "how big should modules be?" is an open question in C++ with modules; I don't know that forcing it at the build system level will allow appropriate exploration of the space to get a good answer.


Even with interface partitions, the cycles can still occur. Here is a minimal reproducer:
(modified from https://github.com/bazelbuild/bazel/pull/22553#issuecomment-3226994833 )

```starlark
cc_library(
    name = "demo",
    module_interfaces = ["foo.cppm", "bar.cppm", "Mod.cppm"],
)
```

```cpp
// Mod.cppm
export module Mod;
```

```cpp
// bar.cppm
export module Mod:bar;
import :foo;
```

```cpp
// foo.cppm
export module Mod:foo;
// import :bar;
```

change the dependency, and rebuild

```
ERROR: /unit_test/test_deps/BUILD.bazel:1:11: in cc_library rule //unit_test/test_deps:demo: cycle in dependency graph:
    configured target: //unit_test/test_deps:demo
    action from: ActionLookupDataN{actionLookupKey=ConfiguredTargetKey{label=//unit_test/test_deps:demo, config=BuildConfigurationKey[dcdb3dcd5587907fe0a81f69b9a6791772e88c9b8ec6c7b857d2e61ff5120fd9]}, actionIndex=11}
.-> action from: ActionLookupData7{actionLookupKey=ConfiguredTargetKey{label=//unit_test/test_deps:demo, config=BuildConfigurationKey[dcdb3dcd5587907fe0a81f69b9a6791772e88c9b8ec6c7b857d2e61ff5120fd9]}, actionIndex=7}
|   action from: ActionLookupData5{actionLookupKey=ConfiguredTargetKey{label=//unit_test/test_deps:demo, config=BuildConfigurationKey[dcdb3dcd5587907fe0a81f69b9a6791772e88c9b8ec6c7b857d2e61ff5120fd9]}, actionIndex=5}
`-- action from: ActionLookupData7{actionLookupKey=ConfiguredTargetKey{label=//unit_test/test_deps:demo, config=BuildConfigurationKey[dcdb3dcd5587907fe0a81f69b9a6791772e88c9b8ec6c7b857d2e61ff5120fd9]}, actionIndex=7}
Target //unit_test/test_deps:demo failed to build
```

fmeum commented:
I think I understand why the non-trivial cycle arises and (roughly) what would need to be done to avoid it. This work seems to be completely independent of this PR though.

@PikachuHyA Since we shouldn't delay this PR any more, what do you think of adding a Starlark check that forces `module_interfaces` to be a singleton (or empty) for now? That would allow us to proceed and get this PR merged while I can look into resolving the underlying cause for the cycle.

mathstuf commented:
Ok, I understand now. I can reproduce the issue with CMake as well. I wonder if `ninja` needs to purge dynamic dependencies from its `.ninja_deps` database when the core-scanned file itself changes.

fmeum commented:
@mathstuf Could you post the relevant output? I would be very interested in learning how this particular problem manifests with other build systems.

PikachuHyA commented:
> @PikachuHyA Could you apply this diff to improve the tests and fix the self cycle?
> 
> Patch

done

PikachuHyA commented:
> @PikachuHyA Since we shouldn't delay this PR any more, what do you think of adding a Starlark check that forces `module_interfaces` to be a singleton (or empty) for now? That would allow us to proceed and get this PR merged while I can look into resolving the underlying cause for the cycle.

I agree. added a check, and deleted the `test_cpp20_modules_with_clang_no_cycle`.

fmeum commented:
@PikachuHyA https://github.com/bazelbuild/bazel/commit/0cfa957c176c952f204feca9ae3c7289f841de1b unfortunately created new merge conflicts

mathstuf commented:
> @mathstuf Could you post the relevant output? I would be very interested in learning how this particular problem manifests with other build systems.

I reported the issue to `ninja` here: https://github.com/ninja-build/ninja/issues/2666

I suspect the solution proposed at the end there also applies here: when loading results from a scan for a module file, drop edges loaded from the cache for that module file (spitballing; I'm not familiar with the implementation details here).

fmeum commented on src/test/shell/bazel/cc_integration_test.sh:
```diff
@@ -2117,4 +2117,87 @@ EOF
+  # Verify that the build can hit the cache without action cycles.
+  bazel clean || fail "Expected clean success"
+  bazel build //:main --experimental_cpp_modules --repo_env=CC=clang --copt=-std=c++20 --disk_cache=disk &> $TEST_log || fail "Expected build C++20 Modules success with compiler 'clang'"
+  expect_log "20 disk cache hit"
```
```suggestion
  expect_log "17 disk cache hit"
```
Something changed at HEAD which affected the total number of actions.

PikachuHyA replied:
done. CI is green now.

fmeum reviewed (commented):
@PikachuHyA One more change and CI should be green

ChuanqiXu9 commented:
@comius @trybka hi, C++20 modules would be usable after this patch. Would you like to take a look. I hope we can use C++20 modules with bazel in 2025.

PikachuHyA commented:
> Synced up with @comius and @pzembrod this morning. I think there will hopefully be a lull in the Java->Starlark migration next week which seems like a good opportunity to try to land this again.
> 
> Can you resolve the current conflicts and I will begin an import on Monday/Tuesday?
> 
> cc: @fmeum

I will update the PR as soon as possible.

PikachuHyA commented:
upgraded the rules_cc and sent https://github.com/bazelbuild/bazel/pull/27139

The macOS CI failed, so I skiped it.

@trybka @comius @fmeum

trybka commented:
Just a quick status update: as you can see there are now a number of merge conflicts. I'm resolving those in my local branch to ease the back and forth. Had to fix a few more tests and the automatic merge resolution had some fun errors.

trybka commented:
I've done my best to resolve merge conflicts but I have a few test failures, @PikachuHyA @fmeum can you take a look?

https://buildkite.com/bazel/google-bazel-presubmit/builds/97272

(should be able to view the current state in https://bazel-review.googlesource.com/c/bazel/+/294090 ? )

fmeum commented:
@trybka With this diff

```patch
--- a/src/main/java/com/google/devtools/build/lib/rules/cpp/CcStarlarkInternal.java
+++ b/src/main/java/com/google/devtools/build/lib/rules/cpp/CcStarlarkInternal.java
@@ -695,6 +695,9 @@ public class CcStarlarkInternal implements StarlarkValue {
       boolean needsIncludeValidation,
       String toolchainType)
       throws EvalException, TypeException {
+    if (modmapInputFile == null || modmapInputFile == Starlark.NONE) {
+      throw new EvalException("modmap_input_file parameter is required");
+    }
     CoptsFilter coptsFilter =
         createCoptsFilter(
             Starlark.isNullOrNone(coptsFilterObject) ? null : (String) coptsFilterObject);
```

I get a stack trace in rules_cc. Is it possible that the same patch also has to be applied to and released with rules_cc? I have to admit that I don't fully understand the split between Bazel builtins and rules_cc anymore.

Edit: The stack trace:
```
ERROR: /private/var/tmp/_bazel_fmeum/507738cfc7e6cde00e4a0230e9aa0722/sandbox/darwin-sandbox/4337/execroot/_main/_tmp/01b593ee1908b4aeba024c8f2007c393/workspace/BUILD.bazel:19:10: in cc_binary rule //:main: 
Traceback (most recent call last):
	File "/private/var/tmp/_bazel_fmeum/507738cfc7e6cde00e4a0230e9aa0722/sandbox/darwin-sandbox/4337/execroot/_main/_tmp/01b593ee1908b4aeba024c8f2007c393/root/5391ae7b4c966eb32ca7d54a9a74ab30/external/rules_cc+/cc/private/rules_impl/cc_binary.bzl", line 813, column 44, in _impl
		binary_info, providers = cc_binary_impl(ctx, [])
	File "/private/var/tmp/_bazel_fmeum/507738cfc7e6cde00e4a0230e9aa0722/sandbox/darwin-sandbox/4337/execroot/_main/_tmp/01b593ee1908b4aeba024c8f2007c393/root/5391ae7b4c966eb32ca7d54a9a74ab30/external/rules_cc+/cc/private/rules_impl/cc_binary.bzl", line 518, column 67, in cc_binary_impl
		(compilation_context, compilation_outputs) = cc_common.compile(
	File "/private/var/tmp/_bazel_fmeum/507738cfc7e6cde00e4a0230e9aa0722/sandbox/darwin-sandbox/4337/execroot/_main/_tmp/01b593ee1908b4aeba024c8f2007c393/root/5391ae7b4c966eb32ca7d54a9a74ab30/external/rules_cc+/cc/private/cc_common.bzl", line 537, column 19, in _compile
		return compile(
	File "/private/var/tmp/_bazel_fmeum/507738cfc7e6cde00e4a0230e9aa0722/sandbox/darwin-sandbox/4337/execroot/_main/_tmp/01b593ee1908b4aeba024c8f2007c393/root/5391ae7b4c966eb32ca7d54a9a74ab30/external/rules_cc+/cc/private/compile/compile.bzl", line 338, column 31, in compile
		_create_cc_compile_actions(
	File "/private/var/tmp/_bazel_fmeum/507738cfc7e6cde00e4a0230e9aa0722/sandbox/darwin-sandbox/4337/execroot/_main/_tmp/01b593ee1908b4aeba024c8f2007c393/root/5391ae7b4c966eb32ca7d54a9a74ab30/external/rules_cc+/cc/private/compile/compile.bzl", line 601, column 53, in _create_cc_compile_actions
		_create_pic_nopic_compile_source_actions(
	File "/private/var/tmp/_bazel_fmeum/507738cfc7e6cde00e4a0230e9aa0722/sandbox/darwin-sandbox/4337/execroot/_main/_tmp/01b593ee1908b4aeba024c8f2007c393/root/5391ae7b4c966eb32ca7d54a9a74ab30/external/rules_cc+/cc/private/compile/compile.bzl", line 821, column 53, in _create_pic_nopic_compile_source_actions
		nopic_object = _create_compile_source_action(
	File "/private/var/tmp/_bazel_fmeum/507738cfc7e6cde00e4a0230e9aa0722/sandbox/darwin-sandbox/4337/execroot/_main/_tmp/01b593ee1908b4aeba024c8f2007c393/root/5391ae7b4c966eb32ca7d54a9a74ab30/external/rules_cc+/cc/private/compile/compile.bzl", line 1018, column 42, in _create_compile_source_action
		_cc_internal.create_cc_compile_action(
Error in create_cc_compile_action: modmap_input_file parameter is required
```

PikachuHyA commented:
> I've done my best to resolve merge conflicts but I have a few test failures, @PikachuHyA @fmeum can you take a look?
> 
> https://buildkite.com/bazel/google-bazel-presubmit/builds/97272
> 
> (should be able to view the current state in https://bazel-review.googlesource.com/c/bazel/+/294090 ? )

@trybka  Thanks for your work. 

I have a basic question. I’m not familiar with Gerrit. How can I check out the change at https://bazel-review.googlesource.com/c/bazel/+/294090 on my development machine, including fetching your updated patchsets? 

Also, regarding the merge conflicts and failing tests: should I resolve the GitHub PR conflicts locally and push the updates to this PR?

trybka commented:
> Also, regarding the merge conflicts and failing tests: should I resolve the GitHub PR conflicts locally and push the updates to this PR?

No, I have a pile of merge conflicts already in my local copy. FWICT the current test failure is just because it's not in rules_cc yet, so we're going to disable it for now. I think that's the only failing test at the moment.

fmeum commented:
Yes, C++20 modules are supported with `last_green` and will be supported by Bazel 9, with the following caveats:
* As of this commit you would still need to use an override to depend on the latest commit of `rules_cc`, which is not part of a release yet.
* Two-phase compilation (the 5/5 part) is not yet supported. I am positive that this can be cherry-picked into a later 9.x minor release.
* Each `cc_library` so far only accepts a single module interface. I will try to land a follow-up change that removes this limitation, but that may also have to wait for one of the 9.x releases.

@PikachuHyA please correct me if I missed anything. And thanks again for the contribution!

PikachuHyA commented:
> For those of us anxiously following from the side lines, does this mean C++ modules are now supported? I'm unsure where the 5/5 of bug resides, or whether this is actually all that's needed (or if this is being rolled out in a downstream bazel release).
> 
> Thank you all for your hard work!

@jwhpryor Thanks for following this work.

With this PR merged, Bazel now offers limited support for C++20 Modules. The current limitation is that each `cc_binary/cc_library`'s `module_interfaces` attribute may contain only a single module interface file (e.g. `foo.cppm`). This limitation can be removed after https://github.com/bazelbuild/bazel/pull/27492 merged (special thanks to @fmeum ). 

The work in https://github.com/bazelbuild/bazel/pull/22555 will not block anyone who wants to use C++20 Modules with Bazel. I will update that PR later. Currently only Clang supports two-phase compilation; GCC/MSVC do not yet.

Our next priorities include broader module features (e.g. support for `import std`), and improving GCC/MSVC support. I'll be working on those items.

lberki commented:
I have to rain on your parade here, I'm afraid: in my capacity as the person whose job this week is to keep our internal builds running, http://www.github.com/bazelbuid/bazel/pull/22553 breaks a few things. I asked @comius to investigate and I hope it's just some tiny little snag, but for now, that's just a hope and I can't exclude the possibility that a significant amount of work will still need to be done (and I don't know yet where, either)
